Tickets/dm 40302

pcortes80 opened the pull request:
- Create the **FactoryTcpServer** class, **createTcpServer.vi**, **Read_currentNumber.vi**, and **Write_currentNumber.vi**.
- Create the `notifierStopServer` to inform the `TcpServer` instances to stop.
- Create the **destroyReferences.vi** to destroy notifier and user event references after the TCP servers are stopped.
- Remove the LabVIEW occurrences to stop the TCP servers.
- Update tests, UML class diagram, and documents.

teweitsai commented on doc/uml/TcpServerClass.uml:
```diff
@@ -1,6 +1,9 @@
 @startuml
+Interface TcpServer
+FactoryTcpServer ..> TcpServer
+TcpServer ..|> TcpServerA
 TcpServerBase <|-- TcpServerCmd
 TcpServerBase <|-- TcpServerTel
-TcpServer *-- TcpServerCmd
-TcpServer *-- TcpServerTel
+TcpServerA *-- TcpServerCmd
+TcpServerA *-- TcpServerTel
```
The class diagram should be:

"""
@startuml
TcpServerBase <|-- TcpServerCmd
TcpServerBase <|-- TcpServerTel
TcpServer *-- TcpServerCmd
TcpServer *-- TcpServerTel
FactoryTcpServer --> TcpServer
@enduml
"""

pcortes80 replied:
Done. Thanks !

teweitsai reviewed (approved):
Looks good. But it looks like there is some problem of resource management. Please discuss first before the modification. Thanks!

[Overview]

1. When you let the TcpServer instances to have the different notifiers, you need to be careful in the releasing part to make sure you do release all the notifier instances.


[Document]

1. You need to update the doc/userManual.md for the screenshot that contains the instantiation of TcpServer instance to use the factory instead.


[FactoryTcpServer.lvclass]

1. You do not have the private vi. Therefore, you can remove the "Private" virtual directory.


[Read_currentNumber.vi]

1. The terminal should not be the "Dynamic Dispatch Input". Use the default one instead.


[Write_currentNumber.vi]

1. The terminal should not be the "Dynamic Dispatch Input". Use the default one instead.


[createTcpServer.vi]

1. Please connect the error line to "Format into String" unless you have a good reason to ignore it.


[createNotifierStopServer.vi]

1. You should not need this vi anymore. Remove it.


[sendNotifierStopServer.vi]

1. This vi should be moved into the TcpServerBase class. In addition, it would be more generic to rename the vi to be stopServer.vi (and more consistent with the meaning in TcpServer.stopServer.vi).


[sendStopServerNotifier.vi]

1. This is a duplication vi. Remove it. By the way, this vi's name is confusing with "sendNotifierStopServer.vi" above.


[waitNotifierStopServer.vi]

1. Both of the TcpServerCmd and TcpServerTel classes have this vi. Therefore, this vi should belong to the TcpServerBase class.
In addition, since there will be two or three processes to call this TcpServerBase.waitNotifierStopServer.vi, maybe we should put it into the preallocated vi. What do you think?


[destroyReferences.vi]

1. This vi calls the "TcpServerBase.releaseNotifierStopServer.vi" and "TcpServerBase.releaseNotifier.vi". I think it would be better to put the functionality of "TcpServerBase.releaseNotifierStopServer.vi" into the "TcpServerBase.releaseNotifier.vi" to decrease the number of vis.

2. In addition, why this "destroyReferences.vi" does not belong to "TcpServerBase" class? How to consider the case that some common notifier might had been removed already? Does the labview ignore this or we can use the "clear error" vi to make sure the error line is still clear?

[configServer.vi]

1. You need to rename the "TcpServerBase.createNotifier.vi" to "TcpServerBase.createNotifierServerStatus.vi". In addition, this vi should be put into the Factory method if you have the similar consideration for multiple instances problem. By doing this, you do not need this "TcpServerBase.createNotifier.vi".


[runServer.vi]

1. I think when this vi is calling the TcpServerCmd.destroyReferences.vi, it does not destroy the related references belong to TcpServerTel class.


[stopServer.vi]

1. You should call "TcpServerBase.stopServer.vi" instead as the comment above.


[waitToStopGetClientStatus.vi]

1. This vi should call the "TcpServerBase.waitNotifierStopServer.vi". Right?


[testCreateTcpServer.vi]

1. You should release the allocated notifier references. Right?


[instantiateTwoTcpServer.vi]

1. The block diagram looks weird. In addition, the order of TcpServer1 and TcpServer2 is reversed.


[serverInstancesTest.vi]

1. You can remove the label of "Fixed in DM-40302".


[testCreateNotifierStopServer.vi]

1. You should not need this vi anymore. Remove it.


[testGetInstanceGenServersAandB.vi]

1. This should release the references of notifiers. Right?


[testStopServersAandB.vi]

1. You can call the instantiateTwoTcpServer.vi or not?

2. There is the overlap of error lines.


[testStopServersAandBMultipleTimes.vi]

1. You can call the instantiateTwoTcpServer.vi or not?

2. You should have the shift register + combine error for the second error line.

pcortes80 commented on doc/userManual.md:
```diff
@@ -15,6 +15,12 @@ The class diagram is in `doc/uml` directory.
+## Factory TcpServer Class
+
+The [Factory Method Pattern](https://en.wikipedia.org/wiki/Factory_method_pattern) is used to instantiate one or more `TcpServer`.
+The `currentNumber` attribute is used as a `TcpServer` ID.
+The **createTcpServer.vi** is a method used to instantiate the `TcpServer` depending on how many times this method is called.
```
Done. Thanks!

pcortes80 commented on src/README.md:
```diff
@@ -39,6 +39,7 @@ labview64 src/buildApp.vi
 There is a class called **TcpServerBase** that is the parent.
 This class has two children: a class called **TcpServerCmd**, to handle commands and events, and a class called **TcpServerTel**, to handle telemetry messages.
 Both children are part of the **TcpServer** class.
+The **FactoryTcpServer** class instantiates one or more **TcpServer** depending on how many times the `createTcpServer` method is called.
```
Done. Thanks!

pcortes80 commented:
[Document]

1. You need to update the doc/userManual.md for the screenshot that contains the instantiation of TcpServer instance to use the factory instead.
**Reply: Done.**

[FactoryTcpServer.lvclass]

1. You do not have the private vi. Therefore, you can remove the "Private" virtual directory.
**Reply: Removed.**

[Read_currentNumber.vi]

1. The terminal should not be the "Dynamic Dispatch Input". Use the default one instead.
**Reply: Fixed.**

[Write_currentNumber.vi]

1. The terminal should not be the "Dynamic Dispatch Input". Use the default one instead.
**Reply: Fixed.**

[createTcpServer.vi]

1. Please connect the error line to "Format into String" unless you have a good reason to ignore it.
**Reply: Done.**

[createNotifierStopServer.vi]

1. You should not need this vi anymore. Remove it.
**Reply: Removed.**

[sendNotifierStopServer.vi]

1. This vi should be moved into the TcpServerBase class. In addition, it would be more generic to rename the vi to be stopServer.vi (and more consistent with the meaning in TcpServer.stopServer.vi).
**Reply: Renamed and moved the vi to the TcpServerBase class as recommended.**

[sendStopServerNotifier.vi]

1. This is a duplication vi. Remove it. By the way, this vi's name is confusing with "sendNotifierStopServer.vi" above.
**Reply: Removed.**

[waitNotifierStopServer.vi]

1. Both of the TcpServerCmd and TcpServerTel classes have this vi. Therefore, this vi should belong to the TcpServerBase class. In addition, since there will be two or three processes to call this TcpServerBase.waitNotifierStopServer.vi, maybe we should put it into the preallocated vi. What do you think?
**Reply: I modified the VI as recommended. Put the VI in preallocated clon execution after the experimentation. Added some notes in src/README.md.**

[destroyReferences.vi]

1. This vi calls the "TcpServerBase.releaseNotifierStopServer.vi" and "TcpServerBase.releaseNotifier.vi". I think it would be better to put the functionality of "TcpServerBase.releaseNotifierStopServer.vi" into the "TcpServerBase.releaseNotifier.vi" to decrease the number of vis.
**Reply: Done.** 

2. In addition, why this "destroyReferences.vi" does not belong to "TcpServerBase" class? How to consider the case that some common notifier might had been removed already? Does the labview ignore this or we can use the "clear error" vi to make sure the error line is still clear?
**Reply: I put the destroyReferences.vi in the TcpServerBase class. Removed the createNotifiers.vi and use the Error Clear function as recommended.**

[configServer.vi]

1. You need to rename the "TcpServerBase.createNotifier.vi" to "TcpServerBase.createNotifierServerStatus.vi". In addition, this vi should be put into the Factory method if you have the similar consideration for multiple instances problem. By doing this, you do not need this "TcpServerBase.createNotifier.vi".
**Reply: I did rename that VI to Tcpserver.createNotifierServerStatus.vi and put it in the "FactoryTcpServer.createTcpServer.vi” to help readability. However, I’m going to change the serverStatusNotifier name to notifierServerStatus in the next ticket.**

[runServer.vi]
I think when this vi is calling the TcpServerCmd.destroyReferences.vi, it does not destroy the related references belong to TcpServerTel class.
**Reply: Done.**

[stopServer.vi]

1. You should call "TcpServerBase.stopServer.vi" instead as the comment above.
**Reply: Done.**

[waitToStopGetClientStatus.vi]

1. This vi should call the "TcpServerBase.waitNotifierStopServer.vi". Right?
**Reply: Done.**

[testCreateTcpServer.vi]

1. You should release the allocated notifier references. Right?
**Reply: Done.**

[instantiateTwoTcpServer.vi]

1. The block diagram looks weird. In addition, the order of TcpServer1 and TcpServer2 is reversed.
Reply: Fixed.

[serverInstancesTest.vi]

1. You can remove the label of "Fixed in DM-40302".
**Reply: Done.**

[testCreateNotifierStopServer.vi]

1. You should not need this vi anymore. Remove it.
**Reply: Done.**

[testGetInstanceGenServersAandB.vi]

1. This should release the references of notifiers. Right?
**Reply: Done.**

[testStopServersAandB.vi]

1. You can call the instantiateTwoTcpServer.vi or not? There is the overlap of error lines.
**Reply: Done.**

[testStopServersAandBMultipleTimes.vi]

1. You can call the instantiateTwoTcpServer.vi or not? You should have the shift register + combine error for the second error line.
**Reply: Done.**

****************************

pcortes80 commented:
After these modifications, I had to modify other test VIs because of resource management in the notifiers. After this improvement now I can run the **testAll.vi** in the correct order, I mean the simple unit test first and then the more complex unit tests.
